kernel/src/actions/mod.rs

```diff
@@ -321,44 +331,84 @@ impl Protocol {
-) -> DeltaResult<()>
+#[inline]
+fn create_feature_error<T>(
+    unsupported: Vec<T>,
+    unsupported_or_unknown: &str,
+    features: &HashSet<T>,
+) -> Error
 where
-    <T as FromStr>::Err: Display,
-    T: Debug + FromStr + Hash + Eq,
+    T: ToString + Debug,
 {
-    let error = |unsupported, unsupported_or_unknown| {
-        let supported = supported_features.iter().collect::<Vec<_>>();
-        let features_type = type_name::<T>()
-            .rsplit("::")
-            .next()
-            .unwrap_or("table features");
-        Error::Unsupported(format!(
-            "{} {} {:?}. Supported {} are {:?}",
-            unsupported_or_unknown, features_type, unsupported, features_type, supported
-        ))
-    };
-    let parsed_features: HashSet<T> = table_features
+    let unsupported = unsupported
         .iter()
-        .map(|s| T::from_str(s).map_err(|_| error(vec![s.to_string()], "Unknown")))
-        .collect::<Result<_, Error>>()?;
+        .map(ToString::to_string)
+        .collect::<Vec<_>>();
+
+    Error::Unsupported(format!(
+        "{} reader features {:?}. Supported reader features are {:?}",
+        unsupported_or_unknown, unsupported, features
+    ))
+}
+
+pub(crate) trait UnknownFeature: Clone + Hash + Eq + Display + Debug {
+    fn is_unknown(&self) -> bool;
+    fn build_unknown(feature: String) -> Self;
+}
+
+impl UnknownFeature for ReaderFeature {
+    fn is_unknown(&self) -> bool {
+        matches!(self, ReaderFeature::Unknown(_))
+    }
+
+    fn build_unknown(feature: String) -> Self {
+        ReaderFeature::Unknown(feature)
+    }
+}
 
-    // check that parsed features are a subset of supported features
+impl UnknownFeature for WriterFeature {
+    fn is_unknown(&self) -> bool {
+        matches!(self, WriterFeature::Unknown(_))
+    }
+
+    fn build_unknown(feature: String) -> Self {
+        WriterFeature::Unknown(feature)
+    }
+}
+
+pub(crate) fn ensure_supported_feature<T: UnknownFeature>(
+    table_features: &[T],
+    supported_features: &HashSet<T>,
+) -> DeltaResult<()> {
+    let (unknown_features, known_features): (Vec<_>, Vec<_>) =
+        table_features.iter().cloned().partition(|f| f.is_unknown());
+    if !unknown_features.is_empty() {
+        return Err(create_feature_error(
+            unknown_features,
+            "Unknown",
+            supported_features,
+        ));
+    }
+    let parsed_features = HashSet::from_iter(known_features);
     parsed_features
         .is_subset(supported_features)
         .then_some(())
         .ok_or_else(|| {
             let unsupported = parsed_features
                 .difference(supported_features)
-                .map(|f| format!("{:?}", f))
+                .cloned()
                 .collect::<Vec<_>>();
-            error(unsupported, "Unsupported")
+            create_feature_error(unsupported, "Unsupported", supported_features)
         })
 }
 
+pub(crate) fn feature_or_unknown<T: UnknownFeature + FromStr>(
```

Comment:
what if instead we have a `TableFeature` trait which both reader and writer features can implement? then we can have a in an `impl TableFeature` a `parse` method that does what you have here?

Reply:
actually do we need any of this at all? can't we just `T::from_str()` inline?

Reply:
It was inline and I thought it looked ugly all inline so I hoisted it up